File: javascript/dsa/algorithms/dp/2d/min-falling-path-sum.js
var minFallingPathSum = function(matrix) {
    const n = matrix.length;
    let dp = Array(n).fill().map(() => Array(n).fill(0)); // @viz: init
    
    for (let j = 0; j < n; j++) { // @viz: baseInit
        dp[0][j] = matrix[0][j]; // @viz: baseInit
    }
    
    for (let i = 1; i < n; i++) { // @viz: loopI
        for (let j = 0; j < n; j++) { // @viz: loopJ
            const mid = dp[i-1][j]; // @viz: computePrep
            const left = j > 0 ? dp[i-1][j-1] : Infinity; // @viz: computePrep
            const right = j < n - 1 ? dp[i-1][j+1] : Infinity; // @viz: computePrep
            
            dp[i][j] = matrix[i][j] + Math.min(mid, left, right); // @viz: compute
        }
    }
    return Math.min(...dp[n-1]); // @viz: return
};

File: cpp/dsa/algorithms/dp/2d/min-falling-path-sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        vector<vector<int>> dp(n, vector<int>(n, 0)); // @viz: init
        for (int j = 0; j < n; j++) { // @viz: baseInit
            dp[0][j] = matrix[0][j]; // @viz: baseInit
        }
        
        for (int i = 1; i < n; i++) { // @viz: loopI
            for (int j = 0; j < n; j++) { // @viz: loopJ
                int mid = dp[i-1][j]; // @viz: computePrep
                int left = (j > 0) ? dp[i-1][j-1] : INT_MAX; // @viz: computePrep
                int right = (j < n - 1) ? dp[i-1][j+1] : INT_MAX; // @viz: computePrep
                
                dp[i][j] = matrix[i][j] + min({mid, left, right}); // @viz: compute
            }
        }
        return *min_element(dp[n-1].begin(), dp[n-1].end()); // @viz: return
    }
};